comps/llms/rayllm/ray_serve.py

```diff
@@ -0,0 +1,525 @@
+# Copyright (c) 2024 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import functools
+import os
+import re
+from enum import Enum
+from queue import Empty
+from typing import Any, AsyncGenerator, Dict, List, Literal, Optional, Tuple, Union
+
+import ray
+import torch
+from fastapi import HTTPException
+from pydantic import BaseModel
+from ray import serve
+from rayllm.api_openai_backend.openai_protocol import ChatMessage, ErrorResponse, ModelResponse
+from rayllm.api_openai_backend.tools import ChatPromptCapture, OpenAIToolsPrompter
+from starlette.requests import Request
+from starlette.responses import JSONResponse, StreamingResponse
+from transformers import AutoModelForCausalLM, AutoTokenizer, TextIteratorStreamer
+
+DEVICE_CPU = "cpu"
+DEVICE_HPU = "hpu"
+
+
+def load_tokenizer(model, tokenizer_name_or_path):
+    tokenizer = AutoTokenizer.from_pretrained(tokenizer_name_or_path)
+    if not model.config.is_encoder_decoder:
+        tokenizer.padding_side = "left"
+    # Some models like GPT2 do not have a PAD token so we have to set it if necessary
+    if model.config.model_type == "llama":
+        # unwind broken decapoda-research config
+        model.generation_config.pad_token_id = 0
+        model.generation_config.bos_token_id = 1
+        model.generation_config.eos_token_id = 2
+        tokenizer.bos_token_id = model.generation_config.bos_token_id
+        tokenizer.eos_token_id = model.generation_config.eos_token_id
+        tokenizer.pad_token_id = model.generation_config.pad_token_id
+        tokenizer.pad_token = tokenizer.decode(tokenizer.pad_token_id)
+        tokenizer.eos_token = tokenizer.decode(tokenizer.eos_token_id)
```

Comment:
We are adopting chat format with huggingface's chat template mechannism, may consider to use that. refer to https://github.com/intel/llm-on-ray/pull/199

Reply:
got it. thanks.